JDK-8298448: UndefinedBehaviorSanitizer

jcking opened the pull request:
Allow building OpenJDK with UBSan. Currently the build fails when optimizing the image due to lots of undefined behavior (it invokes the built JVM). Follow up PRs will either replace the undefined behavior with well defined behavior or suppress errors which are intentional. The goal is to make OpenJDK more well defined and thus more portable across compilers and architectures.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8298448](https://bugs.openjdk.org/browse/JDK-8298448): UndefinedBehaviorSanitizer


### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [d878ede0](https://git.openjdk.org/jdk/pull/11604/files/d878ede08794f3e304582788093e80f3187e531e)
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11604/head:pull/11604` \
`$ git checkout pull/11604`

Update a local copy of the PR: \
`$ git checkout pull/11604` \
`$ git pull https://git.openjdk.org/jdk pull/11604/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11604`

View PR using the GUI difftool: \
`$ git pr show -t 11604`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11604.diff">https://git.openjdk.org/jdk/pull/11604.diff</a>

</details>

robehn commented:
Two questions, this works fine for me:
`UBSAN_CFLAGS="-fsanitize=undefined -fsanitize=float-divide-by-zero -Wno-stringop-truncation -Wno-format-overflow -fno-omit-frame-pointer" ; UBSAN_LDFLAGS="-fsanitize=undefined -fsanitize=float-divide-by-zero" ; sh configure --with-conf-name=udf --with-debug-level=fastdebug --with-native-debug-symbols=internal -with-version-opt=releaserobbin --with-gtest=/home/rehn/source/gtest181/ --disable-precompiled-headers --enable-dtrace --with-extra-cflags="$UBSAN_CFLAGS" --with-extra-cxxflags="$UBSAN_CFLAGS" --with-extra-ldflags="$UBSAN_LDFLAGS" && make images CONF=udf`
I.e.build works ?

Secondly, the method __ubsan_default_options is in the java launcher.
libjvm.so is a "standalone product" that can be used by different launchers.
So I'm not sure it should be in launcher if it was needed.

jcking commented:
What version of GCC are you using? I'm on 12.2 which is the most recent. It
might be that I'm on a newer build with more UBSan coverage.

__ubsan_default_options is called extremely early before main, during
static initializers. IIRC the launcher uses dynamic loading to get
libjvm.so, so that won't work. But maybe I'm misremembering, I'll double
check.

On Fri, Dec 9, 2022, 6:56 PM Robbin Ehn ***@***.***> wrote:

> Two questions, this works fine for me:
> UBSAN_CFLAGS="-fsanitize=undefined -fsanitize=float-divide-by-zero
> -Wno-stringop-truncation -Wno-format-overflow -fno-omit-frame-pointer" ;
> UBSAN_LDFLAGS="-fsanitize=undefined -fsanitize=float-divide-by-zero" ; sh
> configure --with-conf-name=udf --with-debug-level=fastdebug
> --with-native-debug-symbols=internal -with-version-opt=releaserobbin
> --with-gtest=/home/rehn/source/gtest181/ --disable-precompiled-headers
> --enable-dtrace --with-extra-cflags="$UBSAN_CFLAGS"
> --with-extra-cxxflags="$UBSAN_CFLAGS" --with-extra-ldflags="$UBSAN_LDFLAGS"
> && make images CONF=udf
> I.e.build works ?
>
> Secondly, the method __ubsan_default_options is in the java launcher.
> libjvm.so is a "standalone product" that can be used by different
> launchers.
> So I'm not sure it should be in launcher if it was needed.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openjdk/jdk/pull/11604#issuecomment-1344309622>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHTURS2GYPXYGIN5D2CVZ3WMMXRFANCNFSM6AAAAAASY75DP4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

robehn commented:
> What version of GCC are you using?

gcc 11.3 with libubsan 11.2

Also it seem to big overlap with -Wcast-align(=strict) for the warnings/errors I see and I do like that warning.
Do you have idea if the coverage are pretty much the same this particular warning?
I.e. is anyone of them better than the other?

erikj79 commented on make/autoconf/jdk-options.m4:
```diff
@@ -445,6 +445,46 @@ AC_DEFUN_ONCE([JDKOPT_SETUP_ADDRESS_SANITIZER],
 
+###############################################################################
+#
+# UndefinedBehaviorSanitizer
```
I think this logic fits better in `flags.m4`, otherwise this looks ok to me.

erikj79 replied:
Ah now I understand that this compiles runtime checks into the product. In that case it does actually fit well into jdk-options.m4, so you can leave it there.

magicus replied:
Well, this function actually does two things -- first it checks if ubsan support should be enabled, and that code really belong here. But then if it i enabled, it adds additional compiler options, including globally silencing some warnings, and that should really reside in flags-cflags.m4. 

But, I see we have this pattern for other optional tooling support like ASan and GCov, so I guess this is fine for now.

I'm really looking forward to a better way of handling flags so we wouldn't have to do it like this.

theRealAph commented:
> > What version of GCC are you using?
> 
> gcc 11.3 with libubsan 11.2
> 
> Also it seem to big overlap with -Wcast-align(=strict) for the warnings/errors I see and I do like that warning. Do you have an idea if the coverage are pretty much the same for this particular warning? I.e. is anyone of them better than the other?

You can't really compare them because Undefined Behaviour is a runtime thing rather than a compile-time thing. Something like `x = *(char*)0;` is legal in a conforming C++ program as long as it's never executed.

I vastly prefer UB Sanitizer because it detects a ton of stuff that static analysis can't. But there's plenty of room for both.

theRealAph commented:
I'm really pleased to see this.

Im working on `https://github.com/openjdk/jdk/pull/10920`, which fixes the last of the null pointer dereferences in x86/AArch64 HotSpot. Once that's done I'm thinking of enabling `-fsanitize=null` by default in debug builds, as long as it's not too slow.

dholmes-ora commented on make/autoconf/spec.gmk.in:
```diff
@@ -455,6 +455,9 @@ ifeq ($(ASAN_ENABLED), yes)
   endif
 endif
 
+# UndefinedBehaviorSanitizer
+UBSAN_ENABLED:=@UBSAN_ENABLED@
```
I don't see anything reading this. ??

dholmes-ora replied:
To be clear there was a reason that `ASAN_ENABLED` was originally added to spec.gmk.in (though there was a suggestion it could/would be removed later), and no reason has been given for `UBSAN_ENABLED` being needed.

jcking replied:
Fair. I'll remove it. My previous comment, that I dropped on the floor. was asking about that basically.

dholmes-ora commented on src/java.base/share/native/launcher/main.c:
```diff
@@ -34,6 +34,14 @@
 #include "jli_util.h"
 #include "jni.h"
 
+#ifdef UNDEFINED_BEHAVIOR_SANITIZER
```
I really do not like having to make source code changes to accommodate these kinds of tools.

jcking replied:
Yeah, it is unfortunate. However there is no other way to actually set the defaults nicely. The other alternative is to use environment variables, but they are easy to forget when invoking the launcher manually.

dholmes-ora replied:
Does the env-var override the src declaration, or the other way around?

jcking replied:
UBSAN_OPTIONS overrides the src declaration. I added to the comment mentioning that.

dholmes-ora reviewed (commented):
I think it requires much broader discussion as to whether OpenJDK is actively seen to endorse these tools. Why these tools? What if there are other tools, should we support them all?

I'm not saying use of these tools may not be useful, but actually incorporating them into OpenJDK is a decision that needs to be made at a higher-level IMO.

jcking commented:
> I think it requires much broader discussion as to whether OpenJDK is actively seen to endorse these tools. Why these tools? What if there are other tools, should we support them all?
> 
> I'm not saying use of these tools may not be useful, but actually incorporating them into OpenJDK is a decision that needs to be made at a higher-level IMO.

The sanitizers are integrated directly with GCC and Clang/LLVM and are used by projects such as the Linux kernel. They are also used by companies such as Facebook and Google, which IIRC maintain some of the largest closed source mono repositories on the planet. As the sanitizers are integrated directly with GCC and Clang/LLVM, they are extremely easy to use (no external dependencies), fast, and have no direct alternatives. An alternative would also need to be integrated with the compilers in order to be at par.

Additionally configuration options for using ASan already exist in OpenJDK, so that ship has kinda sailed.

If we feel strongly about a discussion, we should probably discuss all the sanitizers as a whole. However that discussion can be done in parallel, as ASan is already used. Just adding the options to OpenJDK does not mean it is endorsed.

jcking commented:
/integrate

dholmes-ora commented:
@jcking this is not ready for integration. You have one review from build team. You have no reviews from core-libs for launcher change. You haven't even bothered to address the comments I made on the actual changes.

dholmes-ora commented on src/java.base/share/native/launcher/main.c:
```diff
@@ -34,6 +34,14 @@
+#ifdef UNDEFINED_BEHAVIOR_SANITIZER
+// Override weak symbol exposed by UBSan to override default options. This is called by UBSan
+// extremely early during library loading, before main is called.
+JNIEXPORT const char* __ubsan_default_options() {
```
Why would this need `JNIEXPORT`? This is not a JNI function.

jcking replied:
Ugh, apparently resolving doesn't add the comment. I thought it did...

So answering, JNIEXPORT is needed so that the symbol is exported and the linker doesn't remove it. It adds __attribute__((visibility("default")).

dholmes-ora replied:
Okay. The limited documentation on these `___xxx_default_options` functions didn't show the need for any export/visibility attributes.

jcking commented:
/reviewers 2

jcking commented:
> @jcking this is not ready for integration. You have one review from build team. You have no reviews from core-libs for launcher change. 

Added reviewers to be 2. I had assumed the bot would take care of getting reviews from both. Apparently it does not, will keep that in mind going forward.

> You haven't even bothered to address the comments I made on the actual changes.

I addressed the comments after integrate, my bad.

jcking commented:
> @jcking this is not ready for integration. You have one review from build team. You have no reviews from core-libs for launcher change. You haven't even bothered to address the comments I made on the actual changes.

Okay, I am dumb. I clicked resolve conversation and assumed it would auto add my comment. Apparently it just drops it and resolves it without saying anything. Sorry!

dholmes-ora commented:
> @dholmes-ora Also etiquette-wise, is it preferred that the commenter resolve the conversation or the author, and the commenter re-open if they feel it is not resolved? I am used to latter workflow, but OpenJDK might have different expectations and I'd like to follow them.

I'm not aware of any specific guidelines. Many people don't even bother with resolving conversations. To me it depends on the nature of the comments as to who should decide if the matter is resolved or not.  It is also unclear of commenting on a "resolved" conversation actually unresolves it again. I don't see an "unresolve" button.

dholmes-ora commented on src/java.base/share/native/launcher/main.c:
```diff
@@ -34,6 +34,14 @@
 #include "jni.h"
 
+#ifdef UNDEFINED_BEHAVIOR_SANITIZER
+// Override weak symbol exposed by UBSan to override default options. This is called by UBSan
```
Please explain why we need to override the default options.

jcking replied:
Done. Also added a comment regarding the environment variable UBSAN_OPTIONS.

dholmes-ora commented on src/java.base/share/native/launcher/main.c:
```diff
@@ -36,7 +36,11 @@
 // Override weak symbol exposed by UBSan to override default options. This is called by UBSan
-// extremely early during library loading, before main is called.
+// extremely early during library loading, before main is called. We need to override the default
+// options because by default UBSan only prints a warning for each occurrence. We want jtreg tests
+// to fail when undefined behavior is encountered. We also want a full stack trace for the offending
```
If this is primarily for tests then can't we set the env-var in the test Makefile?

jcking replied:
Primarily, but its not a requirement. We should also be able to invoke `java` as is. The environment variables should be used to force specific behavior for a single invocation. Otherwise, if one forgets to pass the environment variables, things may not go as expected.

dholmes-ora replied:
It really comes down to who we expect to do these ubsan builds and how we expect them to be used. If this is something explicit then using the env-var seems fine to me. I'm not sure why it needs to be enabled by default through this source modification.

I'd like to hear other opinions on this.

jcking replied:
Sure. It should be noted that whatever we do here for default options will likely have to be repeated for LSan and ASan eventually. So it should be repeatable and easy to maintain. IMO we should enforce strict defaults and then let manual invocations make it less strict via environment variables.

robehn commented on src/java.base/share/native/launcher/main.c:
```diff
@@ -34,6 +34,18 @@
+}
+#endif // UNDEFINED_BEHAVIOR_SANITIZER
+
 /*
```
As I said we have more launcher than 'java', if you put this method here you must put it in all launchers.
I.e. all binaries the call JNI_CreateJavaVM, such out tests, such as jni/daemonDestroy/TestDaemonDestroy.java

jcking replied:
Attempting an approach that automatically includes `__ubsan_default_options` in binaries by "automagically" including a source file for anything using `SetupNativeCompilation` with `TYPE` being `EXECUTABLE`.

jcking replied:
Okay, I was able to get the autoconf magic working. Instead of having to copy-paste `__ubsan_default_options` to every launcher, it is instead auto-inserted in `SetupNativeCompilation` by including a source file which exports it for anything that is `EXECUTABLE`.

jcking commented:
Moving back to draft while I figure out some autoconf magic.

dholmes-ora reviewed (commented):
I like this approach better in principle. NOt sure about the location of the file - the build folk will no doubt have more to say on that. Can we not just have a single file and use:
```
#ifdef __cplusplus
extern "C" {
#endif
```
?
Thanks.

jcking commented:
I tried to use a single file, but the build logic attempts to compile as either C or C++ based on file extensions, and has logic based on it. So if I use `.cpp` and the target is all `.c` odd things happen. The same for the inverse.

erikj79 commented:
> I tried to use a single file, but the build logic attempts to compile as either C or C++ based on file extensions, and has logic based on it. So if I use `.cpp` and the target is all `.c` odd things happen. The same for the inverse.

What happens if you just always use a `.c` file for this? Won't the linker accept that?

jcking commented:
Nope. Some targets end up passing C++ flags to the C compiler, causing a
failure.

On Tue, Dec 13, 2022, 7:13 PM Erik Joelsson ***@***.***>
wrote:

> I tried to use a single file, but the build logic attempts to compile as
> either C or C++ based on file extensions, and has logic based on it. So if
> I use .cpp and the target is all .c odd things happen. The same for the
> inverse.
>
> What happens if you just always use a .c file for this? Won't the linker
> accept that?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openjdk/jdk/pull/11604#issuecomment-1348591254>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHTURUZMU2TVYAD6DH5N63WNB4QRANCNFSM6AAAAAASY75DP4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

theRealAph commented:
I guess the advantage to putting this in the build machinery (as opposed to using
`--with-extra-cflags=-fsanitize=undefined --with-extra-ldflags=-fsanitize=undefined`) is that we can turn some of these onn by default once we've fixed each category of UB. Is that right?

jcking commented:
Correct. We will be able to turn it on by default and enforce it, if w so
choose, for everything.

On Tue, Dec 13, 2022, 8:01 PM Andrew Haley ***@***.***> wrote:

> I guess the advantage to putting this in the build machinery (as opposed
> to using
> --with-extra-cflags=-fsanitize=undefined
> --with-extra-ldflags=-fsanitize=undefined) is that we can turn some of
> these onn by default once we've fixed each category of UB. Is that right?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openjdk/jdk/pull/11604#issuecomment-1348688955>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHTURUVY3ID7BA6BF77WZ3WNCCCVANCNFSM6AAAAAASY75DP4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

robehn commented:
> I guess the advantage to putting this in the build machinery (as opposed to using `--with-extra-cflags=-fsanitize=undefined --with-extra-ldflags=-fsanitize=undefined`) is that we can turn some of these onn by default once we've fixed each category of UB. Is that right?

It will take a while, look a bit on align issue, we have so much code which go from pointer to large -> small -> large, e.g.
```
    address addr = data() + offset;
    return (ImmutableOopMap*) addr;
```
In this case data() needs to return something with the same alignment as a ptr and offset must even in ptr steps.

erikj79 commented:
> Nope. Some targets end up passing C++ flags to the C compiler, causing a failure.

Ah right, we (mis)use CFLAGS (instead of CXXFLAGS) in some SetupNativeCompilation calls when all source files are C++. In that case, your suggested patch makes sense to me.

jcking commented:
> > I guess the advantage to putting this in the build machinery (as opposed to using `--with-extra-cflags=-fsanitize=undefined --with-extra-ldflags=-fsanitize=undefined`) is that we can turn some of these onn by default once we've fixed each category of UB. Is that right?
> 
> It will take a while, look a bit on align issue, we have so much code which go from pointer to large -> small -> large, e.g.
> 
> ```
>     address addr = data() + offset;
>     return (ImmutableOopMap*) addr;
> ```
> 
> In this case data() needs to return something with the same alignment as a ptr and offset must be in even in ptr steps.

Yeah, some of the cases we may just have to suppress if they are not feasible to fix without lots of effort. My intention is to fix all the low hanging fruit, and then suppress the remaining cases. For the suppressed cases, I'll file P4 bugs for each one. Then if somebody feels like fixing them, they can do it.

magicus commented:
Sorry for the late review.

"Magic" solutions like this is always a bit scary. But we have some precedence in terms of the "autoheaders" for assembly files, so I guess this is not worse.

However, I do think the included source files should be treated like the autoheaders, and reside in data rather than in `src`. The latter is intended for buildtools, even though they are a bit scattered at the moment (there is a long-standing JBS bug about fixing that), and these are not really intended to be compiled into a product by itself, rather as something "injected" into all builds. I'd recommend `make/data/ubsan`, and perhaps skip the leading `__`.

magicus commented:
I much also check: is this really needed for *all* executables we make? I would have guessed that it would suffice with the "launchers", i.e. like `java`, `javac`, `jar` etc. These are all compiled from a single source file, `src/java.base/share/native/launcher/main.c`, with different defines depending on what Java classes it should actually launch, and with different output names.

Doing things this way will also affect non-launcher executables, like `jabswitch`, `msiwrapper` and all executable test binaries. 

That might be correct, but I could not be certain of that from trying to read the backlog of discussion in this bug.

jcking commented:
> However, I do think the included source files should be treated like the autoheaders, and reside in data rather than in `src`. The latter is intended for buildtools, even though they are a bit scattered at the moment (there is a long-standing JBS bug about fixing that), and these are not really intended to be compiled into a product by itself, rather as something "injected" into all builds. I'd recommend `make/data/ubsan`, and perhaps skip the leading `__`.

Moved to the directory suggested.

jcking commented:
> I much also check: is this really needed for _all_ executables we make? I would have guessed that it would suffice with the "launchers", i.e. like `java`, `javac`, `jar` etc. These are all compiled from a single source file, `src/java.base/share/native/launcher/main.c`, with different defines depending on what Java classes it should actually launch, and with different output names.
> 
> Doing things this way will also affect non-launcher executables, like `jabswitch`, `msiwrapper` and all executable test binaries.
> 
> That might be correct, but I could not be certain of that from trying to read the backlog of discussion in this bug.

For UBSan, the inclusion in every executable is intentional. For upcoming LSan/ASan stuff, it will be just for launchers. The reason is that UBSan is relaxed by default and won't terminate upon undefined behavior. LSan/ASan are strict by default, and for LSan to work we need to disable some things only when the JVM is being used. So for this commit all executables is correct. For ASan/LSan as you suggest only launchers will have them in that case.

jcking commented:
Friendly poke.

robehn commented:
I think @magicus is back next week. Since this is mainly build stuff I'll sit this one out.

jcking commented:
Ah okay, no problem. Thanks for the update.

magicus commented:
Hi, back from vacation today. Thanks for waiting on my input! This looks good.

jcking commented:
/integrate

magicus commented:
/sponsor
